File: truevfs-kernel/src/main/java/de/schlichtherle/truevfs/kernel/se/ClutchOutputSocket.java
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truevfs.kernel.se;

import java.io.IOException;
import java.io.OutputStream;
import java.nio.channels.SeekableByteChannel;
import javax.annotation.concurrent.NotThreadSafe;
import net.truevfs.kernel.cio.DecoratingOutputSocket;
import net.truevfs.kernel.cio.Entry;
import net.truevfs.kernel.cio.OutputSocket;

/**
 * An output lazySocket which obtains its delegate lazySocket lazily and
 * {@link #reset()}s it upon any {@link Throwable}.
 *
 * @see    ClutchInputSocket
 * @param  <E> the type of the {@link #localTarget() local target}.
 * @author Christian Schlichtherle
 */
@NotThreadSafe
abstract class ClutchOutputSocket<E extends Entry>
extends DecoratingOutputSocket<E> {

    @Override
    protected final OutputSocket<? extends E> socket() throws IOException {
        final OutputSocket<? extends E> socket = this.socket;
        return null != socket ? socket : (this.socket = lazySocket());
    };

    /**
     * Returns the output lazySocket for lazy initialization.
     * 
     * @return the output lazySocket for lazy initialization.
     * @throws IOException on any I/O error. 
     */
    protected abstract OutputSocket<? extends E> lazySocket()
    throws IOException;

    @Override
    public E localTarget() throws IOException {
        try {
            return boundSocket().localTarget();
        } catch (final Throwable ex) {
            reset();
            throw ex;
        }
    }

    @Override
    public SeekableByteChannel channel()
    throws IOException {
        try {
            return boundSocket().channel();
        } catch (final Throwable ex) {
            reset();
            throw ex;
        }
    }

    @Override
    public OutputStream stream() throws IOException {
        try {
            return boundSocket().stream();
        } catch (final Throwable ex) {
            reset();
            throw ex;
        }
    }

    protected final void reset() {
        this.socket = null;
    }
}

File: truevfs-kernel/src/main/scala/de/schlichtherle/truevfs/kernel/se/ClutchOutputSocket.scala
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truevfs.kernel.se

import javax.annotation.concurrent._
import net.truevfs.kernel.cio._
import scala.annotation.tailrec

/** An output socket which obtains its delegate socket lazily and `reset`s
  * it upon any [[java.lang.Throwable]].
  *
  * @see    ClutchInputSocket
  * @tparam E the type of the `localTarget`.
  * @author Christian Schlichtherle
  */
@NotThreadSafe
private abstract class ClutchOutputSocket[E <: Entry]
extends DelegatingOutputSocket[E] {

  private[this] var _socket: Option[OutputSocket[_ <: E]] = None

  override def localTarget() = apply(_.localTarget())

  override def stream() = apply(_.stream())

  override def channel() = apply(_.channel())

  private def apply[A](f: OutputSocket[_ <: E] => A) = {
    try { f(boundSocket) }
    catch { case ex => reset(); throw ex }
  }

  protected final def reset() { _socket = None }

  @tailrec
  protected final override def socket: OutputSocket[_ <: E] = {
    _socket match {
      case Some(socket) => socket
      case None => _socket = Some(lazySocket); socket
    }
  }

  /** Returns the output socket for lazy initialization.
    * 
    * @return the output socket for lazy initialization.
    * @throws IOException on any I/O error. 
    */
  protected def lazySocket: OutputSocket[_ <: E]

  /** Returns a string representation of this object for debugging and logging
    * purposes.
    */
  override def toString = "%s[socket=%s]" format (getClass.getName, socket)
}
